exist-core/src/main/java/org/exist/backup/restore/AppRestoreUtils.java

```diff
@@ -0,0 +1,180 @@
+                                            LOG.warn("Invalid package descriptor for " + app.getSymbolicPath());
+                                            return;
+                                        }
+                                        try {
+                                            final AppDetail detail = new AppDetail(app.getSymbolicPath(), name, Semver.parse(version));
+                                            result.add(detail);
+                                        } catch (PackageException e) {
+                                            LOG.warn("Invalid semver found while parsing " + app.getSymbolicPath());
+                                        }
+                                    }
+                                }
+                            });
+                            reader.parse(is);
+                        } catch (IOException | SAXException e) {
+                            LOG.warn("Parse exception while parsing " + app.getSymbolicPath("expath-pkg.xml", false));
+                        } finally {
+                            if (reader != null) {
+                                parserPool.returnXMLReader(reader);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        return result;
+    }
+
+    private static List<String> getSubcollectionNames(XMLReaderPool parserPool, BackupDescriptor apps) {
+        final List<String> collections = new ArrayList<>();
+        try {
+            apps.parse(parserPool, new DefaultHandler() {
+                @Override
+                public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
+                    if (Namespaces.EXIST_NS.equals(uri) && "subcollection".equals(localName)) {
+                        collections.add(attributes.getValue("filename"));
+                    }
+                }
+            });
+        } catch (IOException | SAXException e) {
+            LOG.warn("SAX error while parsing backup descriptor " + apps.getSymbolicPath(), e);
+        }
+        return collections;
+    }
+
+    /**
+     * Get the database root path for applications, removing /db and trailing slash.
+     * @param broker the broker to get the configuration from
+     * @return the root path for applications
+     */
+    private static String getAppRoot(DBBroker broker) {
+        String appRoot = (String) broker.getConfiguration().getProperty(Deployment.PROPERTY_APP_ROOT);
+        if (appRoot.startsWith(XmldbURI.ROOT_COLLECTION + '/')) {
+            appRoot = appRoot.substring(XmldbURI.ROOT_COLLECTION.length() + 1);
+        }
+        if (appRoot.endsWith("/")) {
+            appRoot = appRoot.substring(0, appRoot.length() - 1);
+        }
+        return appRoot;
+    }
+
+    final static class AppDetail {
+        protected final String path;
```

Comment:
Why `protected`?

Reply:
Just to make codacity stop getting on my nerves ;-)